Support environment variable, path ends with / and determining the right NDK version.

dalinaum opened the pull request:
@realm/java I made a simple patch to supports `ANDROID_NDK`. Please help to review.

emanuelez commented:
In the new build system I'm writing I move away from local.properties and rely on ANDROID_HOME (for the SDK) and NDK_HOME (for the NDK). I think we should align on the name of the environment variables. While ANDROID_HOME is pretty much the standard for the SDK there isn't one for the NDK.
I chose NDK_HOME, but maybe ANDROID_NDK is better since it's easier to spot it's related to Android and will appear close to ANDROID_HOME if you run `env | sort`

cmelchior commented:
+1 for ANDROID_NDK (perhaps ANDROID_NDK_HOME?)

emanuelez commented:
@cmelchior Don't go Cocoa on us :)

dalinaum commented:
How about supporting both `ANDROID_HOME` and `NDK_HOME` all?

beeender commented:
Is it possible to check `RELEASE.TXT` to determine the right ndk version? There is a problem to compile Realm with r10e .

dalinaum commented:
@beeender It is possible. Shoud I add another commit to detetmine the right NDK version?

beeender commented:
I would like to have it. @emanuelez Your opinion?

dalinaum commented:
I added supporting NDK_HOME and checking RELEASE.TXT temporarily. I'll modify this branch as per out futher discussion.

dalinaum commented:
Review please @realm/java

cmelchior commented on realm-jni/build.gradle:
```diff
@@ -87,15 +87,34 @@ if (project.forceDownloadCore instanceof String) {
+        throw new GradleException('The ndk.dir property in the local.properties file and ANDROID_NDK are not set. One of them should be set.')
+    } else if (!file(getNdkDir()).directory) {
+        throw new GradleException('The path provided in the ndk.dir property in the local.properties or ANDROID_NDK is not a folder.')
+    } else if (!file(new File(getNdkDir(), 'RELEASE.TXT')).file) {
```
This check seems really strange? It is just checking if a file `RELEASE.TXT` exists, which doesn't mean if it is a ndkdir or not. Right now just calling `getNdkDir()` will make that check`

I would split `getNdkDir` into two methods `getNdkDir` and `verifyNdkDir` that would make the code and these if-statements more readable IMO

emanuelez replied:
Also you mix Java's `File` with Gradle's `project.file()`. I knowyou copy-pasted my code, but I know better now :yum:

cmelchior commented on realm-jni/build.gradle:
```diff
@@ -87,15 +87,34 @@ if (project.forceDownloadCore instanceof String) {
+    def ndkDir = null
+    if (project.ext.properties.containsKey('ndk.dir')) {
+        ndkDir = project.ext['ndk.dir']
+    } else if (System.getenv('ANDROID_NDK')) {
```
IMO it is confusing to support both. Just pick one of them. You can battle it out with @emanuelez which one you prefer, personally I would vote for `ANDROID_NDK` but I don't have a strong preference.

dalinaum replied:
I don't have a strong preference too. Although other app I installed needs `ANDROID_NDK`, I can set both `ANDROID_NDK` and `NDK_HOME` in my laptop.

cmelchior commented:
Review done. I think it could use a bit of refactoring to make the code a bit easier to read. 
Also we should probably also document this behavior in the `README.md` file

dalinaum commented:
I just pushed a new commit. I chose `NDK_HOME` for the NDK, updated `README.md` and added `verifyNdk`.

cmelchior commented on realm-jni/build.gradle:
```diff
@@ -87,16 +87,39 @@ if (project.forceDownloadCore instanceof String) {
+    if (!ndkDir) {
+        throw new GradleException('The ndk.dir property in the local.properties file and NDK_HOME are not set. One of them should be set.')
+    }
+    if (!file(ndkDir).directory) {
```
Shouldn't this check be part of `getNdkDir`?

dalinaum replied:
Do you mean only `!file(ndkDir).directory` should be part of `getNdkDir`? `!ndkDir` should be part of `getNdk` too?

cmelchior replied:
I mean the entire `if`. This checks that the configured ndkDir actually a folder, so it belongs in `getNdkDir()` IMO.

```
if (!file(ndkDir).directory) {
        throw new GradleException('The path provided in the ndk.dir property in the local.properties or NDK_HOME is not a folder.')
     }
```

cmelchior commented on realm-jni/build.gradle:
```diff
@@ -87,16 +87,39 @@ if (project.forceDownloadCore instanceof String) {
+        throw new GradleException('The path provided in the ndk.dir property in the local.properties or NDK_HOME does not seem to be an Android NDK.')
+    }
+    def ndkVersion = releaseFile.getText('UTF-8').split()[0]
+    if (ndkVersion != 'r10d') {
```
It would be easier to maintain if `r10d` was moved to a constant at the top of the file.

cmelchior commented on realm-jni/build.gradle:
```diff
@@ -87,16 +87,39 @@ if (project.forceDownloadCore instanceof String) {
+    }
+    def ndkVersion = releaseFile.getText('UTF-8').split()[0]
+    if (ndkVersion != 'r10d') {
+        throw new GradleException("Your NDK version: ${ndkVersion}. Realm JNI should be compiled with the version r10d of NDK.")
```
Use constant from above

dalinaum replied:
I just added `gradle.properties` file for this.

cmelchior commented on realm-jni/build.gradle:
```diff
@@ -87,16 +87,39 @@ if (project.forceDownloadCore instanceof String) {
+}
+
+task checkProperties(group: 'check', description: 'Check the user provided gradle.properties') << {
+    verifyNdk()
```
If you instead did:

```
ndkFolder = getNdkDir();
verifyNdk(ndkFolder)
```

You could avoid calling getNdkDir() twice which is happening right now.

dalinaum replied:
I modify this code. But `getNdkDir` is not called twice in this section. `verifyNdk` calls `getNdkDir` once.

cmelchior replied:
Yes, but then you call `getNdkDir` a few lines further down: 

```
"${getNdkDir()}/build/tools/make-standalone-toolchain.sh",
```

dalinaum replied:
It's in generateNdkToolchain${toolchain.name.capitalize()}" tasks, not in checkProperties task. I think it's not in the same scope. If we hope avoid calling `getNdkDir()` there, `ndkFolder` should be a global variable.

cmelchior replied:
Ah yes, sorry. The codeview on GitHub must have tricked me.

dalinaum commented:
Please review again, @realm/java

emanuelez commented on realm-jni/build.gradle:
```diff
@@ -87,16 +87,39 @@ if (project.forceDownloadCore instanceof String) {
+}
+
+def verifyNdk(ndkDir) {
+    def releaseFile = new File("${ndkDir}/RELEASE.TXT")
```
new File( -> file(

dalinaum replied:
Thanks. I just applied this.

dalinaum commented:
How can I solve this? 

```
io.realm.RealmTest > testQueriesFailWithInvalidDataTypes[Nexus 9 - 5.0.1] [31mFAILED [0m
	java.lang.IllegalStateException: It's not allowed to delete the file associated with an open Realm. Remember to close() all the instances of the Realm before deleting its file.
	at io.realm.BaseRealm.deleteRealm(BaseRealm.java:518)
:realm:connectedAndroidTestDebug FAILED
```

It seems not relevant to this PR.

cmelchior commented:
It is because a unit test doesn't properly close it's realm after being done. In the setup method we delete the Realm file to start from scratch, but if a previous unit test didn't close it properly that will fail.

emanuelez commented:
I think I actually found the problem and fixed it in my build system rewrite branch. Let me take a look

dalinaum commented:
I justed repushed the squashed commit, please review again. @realm/java

cmelchior commented:
:+1:

emanuelez commented:
:+1:

kneth commented on README.md:
````diff
@@ -57,6 +56,12 @@ Prerequisites:
+* Or, if you would like to add environment variables to your profile:
+
+    ```
+    export ANDROID_HOME=/Applications/Android Studio.app/sdk
````
This is `bash` (or `zsh`) style.

cmelchior replied:
So you just want the `export` part removed?

kneth replied:
Just forget my comment (we probably don't have many `tcsh` users).

dalinaum replied:
OK. :)

zaki50 replied:
@dalinaum path includes space, so you need quote by "

kneth commented on realm-jni/build.gradle:
```diff
@@ -87,16 +87,39 @@ if (project.forceDownloadCore instanceof String) {
+            ndkDir = ndkDir[0..-2]
+        }
+    } else {
+        throw new GradleException('The ndk.dir property in the local.properties file and NDK_HOME are not set. One of them should be set.')
```
"and" -> "or"
"are" -> "is"

dalinaum replied:
I just fixed them. Thanks.

kneth commented:
:+1: besides my minor comments.

dalinaum commented:
Please review again @realm/java I hope this commit will be the final modification of this PR. :)

zaki50 commented:
:+1:

zaki50 commented:
@dalinaum I added comment for small fix

dalinaum commented:
You're right. I should fix it. I hope this old fashion path will be
replaced with the new school path.
2015. 10. 9. 오후 11:42에 "Makoto Yamazaki" <notifications@github.com>님이 작성:

> @dalinaum <https://github.com/dalinaum> I added comment for small fix
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/realm/realm-java/pull/1539#issuecomment-146891735>.
>

dalinaum commented:
sdk.dir=/<your home directory>/Library/Android/sdkndk.dir=/usr/local/Cellar/android-ndk/r10d
     ```
+* Or, if you would like to add environment variables to your profile:
+
+    ```
+    export ANDROID_HOME=~/Library/Android/sdk
+    export NDK_HOME=/usr/local/Cellar/android-ndk/r10d

zaki50 commented on realm-jni/build.gradle:
```diff
@@ -87,16 +87,39 @@ if (project.forceDownloadCore instanceof String) {
-        throw new GradleException('The path provided in the ndkDir property in the gradle.properties does not seem to be an Android NDK.')
+def getNdkDir() {
+    def ndkDir = null
+    if (project.ext.properties.containsKey('ndk.dir')) {
```
@dalinaum we can simplify `project.ext.properties.containsKey` by `project.hasProperty`.

dalinaum replied:
Thanks @zaki50 for your review. It's 2AM KST and JST :) I am currently rebuild Realm to test `project.hasProperty`. I then will upload the new commit.

zaki50 replied:
@dalinaum I should comment this before last squash. sorry for that.

dalinaum replied:
@zaki50 Never mind. Thanks.

zaki50 commented:
@dalinaum :+1: besides one minor comment

cmelchior commented:
:+1:

dalinaum commented:
Thanks all for reviews and being patient. :)
